### ISSUE TITLE
API Return Error - Attribute Error 'NoneType' object has no attribute 'get' 

### ISSUE THREAD

**Issue by formatBCE** [webdjoe/pyvesync#83]
Hello!
I'm working on VeSync integration for Home Assistant now. Extending it for usage of Classic300 humidifier.
I am able to create humidifier entity, works normal.
However, this humidifier contains also night light and another option of turning display on/off.
I tried to create different entities for these functions (dimmable light for night light, and switch for display). Overall, it works. However, integration config is inconsistent - i can't get all entities registered. Sometimes it succeeds, but 9 of 10 times it gives following error:
`File "/usr/local/lib/python3.8/site-packages/pyvesync/vesyncfan.py", line 757, in update
    self.get_details()
  File "/usr/local/lib/python3.8/site-packages/pyvesync/vesyncfan.py", line 738, in get_details
    outer_result = r.get('result', {})
AttributeError: 'NoneType' object has no attribute 'get'`

And as result, i'm getting only one or two of entities registered in Home Assistant.
My guess is, that VeSync API is rejecting simultaneous calls to same device.
I have also 2 of Levoit 131s purifiers - they're recognized each time correctly.

**Comment by webdjoe** [comment]
The api call is not returning a device response. Possibly a rate limiting
issue or change in API. The best bet would be to run a test script to call
‘device.update()’ several times in a row and record the responses to see
what’s going on. I don’t have a humidifier to test, or I would try it
myself.

**Comment by formatBCE** [comment]
So i ran small script, that creates manager, updates it, and then makes 10 calls for specific device "update" function.
Results are following:
- for 131s it succeeds 10/10 all the time:
`
> python update_check.py
Update  1
Success
Update  2
Success
Update  3
Success
Update  4
Success
Update  5
Success
Update  6
Success
Update  7
Success
Update  8
Success
Update  9
Success
Update  10
Success
`

- for Classic300s it fails all the time, after 2-5 success responses:
`
> python update_check.py
Update %d 1
Success
Update %d 2
Success
Update %d 3
Success
Update %d 4
Success
Update %d 5
2021-06-17 14:26:13,721 - WARNING - HTTPSConnectionPool(host='smartapi.vesync.com', port=443): Read timed out. (read timeout=5)
Traceback (most recent call last):
  File "update_check.py", line 9, in <module>
    humi.update()
  File "...\pyvesync\vesyncfan.py", line 757, in update
    self.get_details()
  File "...\pyvesync\vesyncfan.py", line 738, in get_details
    outer_result = r.get('result', {})
AttributeError: 'NoneType' object has no attribute 'get'
`

I guess i saw somewhere in logs, that for this device different API endpoint is used. Am i right?
Could it be, that this API has some DDoS check, or something?

**Comment by webdjoe** [comment]
Does the call succeed after failing for a certain amount of time? If so, it
probably is rate limiting. Rest apis usually have leaky bucket rate
limiting. It allowed you to make 4 calls so maybe try to sleep 16 seconds
and if that doesn’t work then 30.  I can include an update interval for
that device but I would like to be sure. The code isn’t async so it would
be blocking.

**Comment by formatBCE** [comment]
It seems to be really inconsistent. Sometimes even manager.update() gives me that error.
I tried with 15 seconds, 30 seconds and 120 seconds between calls.
None succeeded, most of runs failing on 2nd-3rd attempt.
So either it's some session-based restriction, or endpoint is too laggy...

**Comment by formatBCE** [comment]
I see, that library does use different API endpoint for 131s.
And this one `/cloud/v2/deviceManaged/bypassV2` is used for both purifier-200 and humidifier-300s. That means, that this issue would affect both of devices, if i had them, right? 
So if particular user has more than one humidifier/purifier device (excluding 131s), he'll have problems adding them...

**Comment by formatBCE** [comment]
Now i guess i got the reason of creation pyvesync_v2 lib.
Does it use new API endpoints?

**Comment by webdjoe** [comment]
Don't use v2, that is obsolete.  I should delete it from pypi. I'll do some
testing on the devices that I have but do you have an android that you
could capture packets on? I want to double check with the original packet
captures I used.

What is the api returning when the error pops up?  Is it a completely empty
object? Is the return code 200?

**Comment by formatBCE** [comment]
I have Android device. It's not rooted. Can i intercept on it?

> What is the api returning when the error pops up? Is it a completely empty object? Is the return code 200?

How do i check that?

**Comment by webdjoe** [comment]
You can capture without rooted android. Just install "Packet Capture" app with the round blue and yellow icon. Here is the link https://play.google.com/store/apps/details?id=app.greyshirts.sslcapture&hl=en_US&gl=US

Please capture device list and device functions. Try to press refresh list and device details quickly to try to mirror the library calls. Export the txt files of the captures and you can email me at webdjoe - at - gmail.com

You don't need to check the return code, I just realized it won't return the response if it is not 200.  There are two options to check response:

1. Debug through the library and look at variables as you step through. If you are using VSCode, you would have to download the library files and use a virtual environment that doesn't have pyvesync installed. Make a test script in the folder with the library files and step through. If you are using PyCharm (totally better) you can run a test script in a virtual env with pyvesync installed and step through all code by pressing F7 (as opposed to just your own code).
2. You can create a test script instantiate the VeSync object and import the Helpers to manually call API. This is just a quick and dirty outline that instantiates and returns a tuple with the response and code. You can put a wrapper in the if name == main section and excute as many times as you would like while output to stdout or wherever:
```python
from pyvesync import VeSync
from pyvesync import Helpers


class VSTester:
      def __init__(self, user, pass):
          self.vs = VeSync(username, password);
          if not self.vs.login():
              raise Exception("Error logging in")
      def device_list(self):          
          response, code = Helpers.call_api(
              '/cloud/v1/deviceManaged/devices',
              'post',
              headers=Helpers.req_headers(self),
              json=Helpers.req_body(self, 'devicelist'),
          )
         return response, code
       def device_info(self):
           self.vs.update()
           fan = self.vs.fans[0] # assuming one fan
           head = Helpers.bypass_header()
           body = Helpers.bypass_body_v2(self.vs)
           body['cid'] = fan.cid
           body['configModule'] = fan.config_module
           body['payload'] = {
               'method': 'getHumidifierStatus',
               'source': 'APP',
               'data': {}
            }

            response, code = Helpers.call_api(
                '/cloud/v2/deviceManaged/bypassV2',
                method='post',
                headers=head,
                json=body,
            )
            return response, code


```

**Comment by formatBCE** [comment]
The "Packet Capture" app struggles to generate certificate, unfortunately. :(

**Comment by formatBCE** [comment]
@webdjoe new info: today that API stopped working.
It is down at least for several hours now.
Not only it affected my custom HA integration, but VeSync app does not work too. Kinda.
It works actually, but only because of dirty trick: this humidifier happened to have BLE also!
I discovered it couple days ago, while was working on some BLE integration in HA. I found the device Classic300s among Bluetooth devices.
So while BT on phone is on, it works. Bluetooth is switched on by the application on start.
So I entered application, and switched Bluetooth off.
Humidifier stopped working immediately.

Looks like, instead of making local LAN, or at least decent API server, they decided to make it WiFi+Bluetooth...

**Comment by nsubtil** [comment]
Saw similar problems, with a call stack suggesting no data was read back and HTTPS timeouts in the log. Bumping API_TIMEOUT to 30 seems to fix this.

**Comment by formatBCE** [comment]
> Saw similar problems, with a call stack suggesting no data was read back and HTTPS timeouts in the log. Bumping API_TIMEOUT to 30 seems to fix this.

From my tracing, errors are returned in different time gaps, doesn't look like timeout...
@webdjoe how could we try to increase timeout inside of pyvesync?

**Comment by formatBCE** [comment]
Today server supposedly got update.
Humidifier got removed from vendor application.
After re-adding, it started to work flawlessly with Alexa, my HA integration code and application itself.
Cool!

**Comment by decompil3d** [comment]
FWIW, I have about a dozen Vesync outlets and just got a 200s air purifier. As soon as I added the air purifier to the system I started seeing these errors frequently in my HA logs and outlets started being very flaky to commands. I just removed the air purifier from my Vesync account today and outlets work well again. So, something is probably wrong with the API communication when the air purifier and humidifier devices are involved.

**Comment by webdjoe** [comment]
Yea, Vesync has been updating their API pretty extensively. I think they are also trying to cut down on people using the api continuously like for HA. They implemented multiple hefty certificate pinning checks in the new android app versions. They designed their system for updates while the app is open but with HA it happens continuously which I'm sure is impacting their server loads. May eventually try to cycle user agents, app versions and device types in case they rate limit the ones that this library uses. The new apps don't use many of the calls that this library uses. I think it's only a matter of time before they are depreciated and they force users to update app. I just pushed a bandaid fix that should gracefully handle timeouts. There's something wrong with azure that is making the checks fail but everything checks out locally. Give it a shot and let me know if this solves it, at least temporarily. #121

**Comment by webdjoe** [comment]
This has been addressed in #121
